### ISSUE TITLE
Replace `content-type` with built-in `util.MIMEType`

### ISSUE THREAD

**Issue by climba03003** [fastify/fastify#4502]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

I see that the core provide a `MIMEType` parser in the recent release which does the same as `content-type` but in more permissive way. For example, `text/plain; application/json; charset=utf-8` is known to be wrong but allowed to parse without problem.

Although, it is slower than `content-type` but combine with the calculation of other property.
I believe it provide more gains.

Benchmark
```js
const util = require('util')
const contentType = require('content-type')
const { parseContentType } = require('busboy/lib/utils')
const Benchmark = require('benchmark');
const suite = new Benchmark.Suite;

const str = 'application/json; charset=utf-8'

suite.add('util#MIMEType', function() {
    new util.MIMEType(str)
  })
  .add('content-type#parse', function() {
    contentType.parse(str)
  })
  .add('busboy#parseContentType', function() {
    parseContentType(str)
  })
  .on('cycle', function(event) {
    console.log(String(event.target));
  })
  .on('complete', function() {
    console.log('Fastest is ' + this.filter('fastest').map('name'));
  })
  .run({ 'async': true });
```

Result

```
util#MIMEType x 779,686 ops/sec ±0.29% (93 runs sampled)
content-type#parse x 877,442 ops/sec ±0.43% (94 runs sampled)
busboy#parseContentType x 541,242 ops/sec ±0.32% (93 runs sampled)
Fastest is content-type#parse
```

**Comment by climba03003** [comment]
The API is relatively new and marked as [`Experimental`](https://nodejs.org/docs/latest-v18.x/api/util.html#class-utilmimetype). So, it is not required to be added on next release.
We can wait until stable and experiment it a bit at the same time.

**Comment by Uzlopak** [comment]
Can I interet you in fast-query-type-parse?

```sh
uzlopak@Battlestation:~/workspace/fast-content-type-parse$ node benchmarks/bench.js
util#MIMEType x 1,117,136 ops/sec ±1.20% (94 runs sampled)
fast-content-type-parse#parse x 3,595,617 ops/sec ±0.71% (91 runs sampled)
fast-content-type-parse#safeParse x 3,573,992 ops/sec ±0.86% (94 runs sampled)
content-type#parse x 1,461,697 ops/sec ±0.56% (91 runs sampled)
busboy#parseContentType x 960,277 ops/sec ±0.37% (93 runs sampled)
Fastest is fast-content-type-parse#parse,fast-content-type-parse#safeParse
```

**Comment by Uzlopak** [comment]
Can I interet you in fast-query-type-parse?

```sh
uzlopak@Battlestation:~/workspace/fast-content-type-parse$ node benchmarks/bench.js
util#MIMEType x 1,117,136 ops/sec ±1.20% (94 runs sampled)
fast-content-type-parse#parse x 3,595,617 ops/sec ±0.71% (91 runs sampled)
fast-content-type-parse#safeParse x 3,573,992 ops/sec ±0.86% (94 runs sampled)
content-type#parse x 1,461,697 ops/sec ±0.56% (91 runs sampled)
busboy#parseContentType x 960,277 ops/sec ±0.37% (93 runs sampled)
Fastest is fast-content-type-parse#parse,fast-content-type-parse#safeParse
```

**Comment by climba03003** [comment]
I am in favor if it can test against https://github.com/web-platform-tests/wpt/tree/master/mimesniff
You may need to modify the test for usage, but I assume the [`resources`](https://github.com/web-platform-tests/wpt/tree/master/mimesniff/mime-types/resources) should be keep untouched.

**Comment by Uzlopak** [comment]
mimesniff seems to be a test suite for sniffing the mimetype based on magic bytes and stuff.

**Comment by climba03003** [comment]
> mimesniff seems to be a test suite for sniffing the mimetype based on magic bytes and stuff.

It also provide the test-suite for verifying if the parser works as intended.

**Comment by Uzlopak** [comment]
What is the goal of this? Do you expect a MIMEType ponyfill for node versions smaller than 19?

**Comment by climba03003** [comment]
> What is the goal of this?

Code simplification.
I think `.essence` will be useful in simplifying the check and it follow how `browser | fetch` parsing the MIMEType. So, we will not have mis-match behavior.

> Do you expect a MIMEType ponyfill for node versions smaller than 19?

If we use `MIMEType`, then the target will be `major` so it will not need any `ponyfill`  or `polyfill` since we drop `14, 16` support at that point.

If we use other alternative, then it can be either `minor` or `major`.

**Comment by Uzlopak** [comment]
tbh. I wanted just supply a faster implementation of the content-type package regarding parsing, which should be a drop in replacement for the current code in fastify. 

If you have in mind, that this should be only accepted if basically a clone of MIMEType is supplied, than I am out.

**Comment by climba03003** [comment]
> .essence is basically what .type is.

No. `.essence` is a loosen check for `content-type`. So, `text/plain; application/json; charset=utf-8` will be permit as valid `content-type` but resolve properly with the first given one `text/plain`.

**Comment by climba03003** [comment]
> tbh. I wanted just supply a faster implementation of the content-type package regarding parsing, which should be a drop in replacement for the current code in fastify.

When you means `drop-in` the behavior shouldn't change. And the `token` allowed is not the same.
[`!#$%&'*+.^_`|~0-9A-Za-z-`](https://github.com/jshttp/content-type/blob/master/index.js#L25) vs [`+._0-9A-Za-z-`](https://www.npmjs.com/package/fast-content-type-parse?activeTab=explore)

**Comment by Uzlopak** [comment]
I investigated the IANA table regarding mimetypes. The characters you are missing were never used.

**Comment by climba03003** [comment]
> I investigated the IANA table regarding mimetypes. The characters you are missing were never used.

But, we cannot restrict it due to no public mimetypes using those characters. Maybe a private one using those strange characters.
Actually, I am open to replace it for performance and see it someone complain on those characters.

**Comment by Uzlopak** [comment]
I benchmarked, it doesnt matter. So I can use the whole pattern.

**Comment by Uzlopak** [comment]
@climba03003 

https://github.com/fastify/fast-content-type-parse

**Comment by mcollina** [comment]
I recommend you open an issue on Node.js with the performance results. I don't think the new mime utilities were optimized at all.

In Fastify v5 we will drop v14 and   v16, and hopefully the new mimetype will be landed in v18, so we could theoretically just use this if the performance is ok.

**Comment by climba03003** [comment]
> I recommend you open an issue on Node.js with the performance results. I don't think the new mime utilities were optimized at all.
> 
> In Fastify v5 we will drop v14 and   v16, and hopefully the new mimetype will be landed in v18, so we could theoretically just use this if the performance is ok.

FYI, it is already in v18. Just released today in v18.13.0 and v19.3.0.

**Comment by Uzlopak** [comment]
Well tbh. I think we just need a small aspect for fastify. So I dont think it makes sense to invest time into MIMEType of node.
